### PR TITLE
[Refactor] CCE

### PR DESCRIPTION
**Pull request by Aloento**
Move all API methods and related structures to a separate file.
Inline all `Extracts` and use the `extract` package instead.
Remove all Builders

### CODE REVIEW

**Comment by Aloento** [comment]
recheck

**Review by artem-lifshits — DISMISSED** [review]

**Comment by Aloento** [comment]
openstack/waf/v1 TestDomainLifecycle `Domain name a.com already exists`

File: openstack/cce/v1/nodes/list.go [hunk 1]
```diff
@@ -0,0 +1,62 @@
...
+
+	"github.com/opentelekomcloud/gophertelekomcloud"
+	"github.com/opentelekomcloud/gophertelekomcloud/internal/extract"
+	"github.com/opentelekomcloud/gophertelekomcloud/openstack"
+)
+
+// List returns collection of nodes.
+func List(client *golangsdk.ServiceClient, clusterID string) (*ListNodes, error) {
+	raw, err := client.Get(fmt.Sprintf("https://%s.%s", clusterID, client.ResourceBaseURL()[8:])+
+		strings.Join([]string{"nodes"}, "/"), nil, openstack.StdRequestOpts())
```
Comment (outcatcher) [inline openstack/cce/v1/nodes/list.go]: There is literally no sense in joining single-element slice :smile: 
Also, generating this URL takes two rows, it's better to move it into variable in this case:
```go
listURL := fmt.Sprintf("https://%s.%s/nodes/", clusterID, client.ResourceBaseURL()[8:]))
raw, err := client.Get(listURL, ...
```

File: openstack/cce/v1/nodes/get.go [hunk 2]
```diff
@@ -1,19 +1,25 @@
...
 
-type ListNodes struct {
-	Kind       string       `json:"kind"`
-	ApiVersion string       `json:"apiVersion"`
-	Metadata   MetadataLink `json:"metadata"`
-	Nodes      []Node       `json:"items"`
-}
+// Get retrieves a particular nodes based on its unique ID and cluster ID.
+func Get(client *golangsdk.ServiceClient, clusterID, k8sName string) (*GetNode, error) {
+	raw, err := client.Get(fmt.Sprintf("https://%s.%s", clusterID, client.ResourceBaseURL()[8:])+
```
Comment (outcatcher) [inline openstack/cce/v1/nodes/get.go]: Base URL for all methods should be moved to a function (I can see such in previous implementation), there is too much duplication right now.

In current implementation joining slice of predefined strings makes little sense and makes code less readable.

File: openstack/cce/v1/nodes/list.go [hunk 3]
```diff
@@ -0,0 +1,62 @@
...
+type Spec struct {
+	ProviderID string  `json:"providerID"`
+	Taints     []Taint `json:"taints"`
+}
+
+type Taint struct {
+	Key    string `json:"key"`
+	Value  string `json:"value"`
+	Effect string `json:"effect"`
+}
```
Comment (outcatcher) [inline openstack/cce/v1/nodes/list.go:61]: Would be nice to have structure documentation

File: openstack/cce/v3/addons/create.go [hunk 4]
```diff
@@ -0,0 +1,64 @@
...
+	Spec RequestSpec `json:"spec" required:"true"`
+}
+
+type CreateMetadata struct {
+	Annotations CreateAnnotations `json:"annotations" required:"true"`
+}
+
+type CreateAnnotations struct {
+	AddonInstallType string `json:"addon.install/type" required:"true"`
+}
```
Comment (outcatcher) [inline openstack/cce/v3/addons/create.go:28]: Documentation recommended.

File: openstack/cce/v3/addons/delete.go [hunk 5]
```diff
@@ -0,0 +1,19 @@
...
+	"fmt"
+	"strings"
+
+	"github.com/opentelekomcloud/gophertelekomcloud"
+)
+
+// Delete will permanently delete a particular addon based on its unique ID.
+func Delete(client *golangsdk.ServiceClient, id, clusterId string) (err error) {
+	_, err = client.Delete(fmt.Sprintf("https://%s.%s", clusterId, client.ResourceBaseURL()[8:])+
+		strings.Join([]string{"addons", id + "?cluster_id=" + clusterId}, "/"), &golangsdk.RequestOpts{
```
Comment (outcatcher) [inline openstack/cce/v3/addons/delete.go]: Nope, joining is an overcomplication here. Either move to a function, as it was before or use `fmt.Sprintf`

File: openstack/cce/v3/addons/get.go [hunk 6]
```diff
@@ -0,0 +1,24 @@
...
+
+	"github.com/opentelekomcloud/gophertelekomcloud"
+	"github.com/opentelekomcloud/gophertelekomcloud/internal/extract"
+)
+
+// Get retrieves a particular addon based on its unique ID.
+func Get(client *golangsdk.ServiceClient, id, clusterId string) (*Addon, error) {
+	raw, err := client.Get(fmt.Sprintf("https://%s.%s", clusterId, client.ResourceBaseURL()[8:])+
+		strings.Join([]string{"addons", id + "?cluster_id=" + clusterId}, "/"), nil, &golangsdk.RequestOpts{
+		OkCodes: []int{200},
```
Comment (outcatcher) [inline openstack/cce/v3/addons/get.go]: Same as previous

File: openstack/cce/v3/addons/list_addon_instances.go [hunk 7]
```diff
@@ -0,0 +1,24 @@
...
+
+	"github.com/opentelekomcloud/gophertelekomcloud"
+	"github.com/opentelekomcloud/gophertelekomcloud/internal/extract"
+)
+
+func ListAddonInstances(client *golangsdk.ServiceClient, clusterID string) (*AddonInstanceList, error) {
+	raw, err := client.Get(fmt.Sprintf("%s?cluster_id=%s",
+		fmt.Sprintf("https://%s.%s", clusterID, client.ResourceBaseURL()[8:])+
+			strings.Join([]string{"addons"}, "/"), clusterID), nil, &golangsdk.RequestOpts{
+		OkCodes: []int{200},
```
Comment (outcatcher) [inline openstack/cce/v3/addons/list_addon_instances.go]: Same as previous.

File: openstack/cce/v3/addons/list_templates.go [hunk 8]
```diff
@@ -0,0 +1,32 @@
...
+func ListTemplates(client *golangsdk.ServiceClient, clusterID string, opts ListOpts) (*AddonTemplateList, error) {
+	q, err := golangsdk.BuildQueryString(opts)
+	if err != nil {
+		return nil, err
+	}
+
+	raw, err := client.Get(fmt.Sprintf("https://%s.%s", clusterID, client.ResourceBaseURL()[8:])+
+		strings.Join([]string{"addontemplates"}, "/")+q.String(), nil, &golangsdk.RequestOpts{
+		OkCodes: []int{200},
+	})
```
Comment (outcatcher) [inline openstack/cce/v3/addons/list_templates.go]: Same as previous.

File: openstack/cce/v1/nodes/update.go [hunk 9]
```diff
@@ -0,0 +1,42 @@
...
+// Update allows nodes to be updated.
+func Update(client *golangsdk.ServiceClient, clusterID, k8sName string, opts UpdateOpts) (*GetNode, error) {
+	b, err := golangsdk.BuildRequestBody(opts, "")
+	if err != nil {
+		return nil, err
+	}
+
+	raw, err := client.Patch(
+		fmt.Sprintf("https://%s.%s", clusterID, client.ResourceBaseURL()[8:])+
+			strings.Join([]string{"nodes", k8sName}, "/"), b, nil, &golangsdk.RequestOpts{
```
Comment (outcatcher) [inline openstack/cce/v1/nodes/update.go]: Same as previous.

File: openstack/cce/v1/nodes/update.go [hunk 10]
```diff
@@ -0,0 +1,42 @@
...
+			OkCodes: []int{200},
+			MoreHeaders: map[string]string{
+				"Content-Type": "application/merge-patch+json",
+			},
```
Comment (outcatcher) [inline openstack/cce/v1/nodes/update.go:35]: Maybe add some kind of `commonCCEHeaders` or `commonCCEOptions` function?

File: openstack/cce/v3/addons/create.go [hunk 11]
```diff
@@ -0,0 +1,64 @@
...
+
+// Create accepts a CreateOpts struct and uses the values to create a new addon.
+func Create(client *golangsdk.ServiceClient, opts CreateOpts, clusterId string) (*Addon, error) {
+	b, err := golangsdk.BuildRequestBody(opts, "")
+	if err != nil {
+		return nil, err
+	}
+
+	raw, err := client.Post(fmt.Sprintf("https://%s.%s", clusterId, client.ResourceBaseURL()[8:])+
+		strings.Join([]string{"addons"}, "/"), b, nil, &golangsdk.RequestOpts{OkCodes: []int{201}})
```
Comment (outcatcher) [inline openstack/cce/v3/addons/create.go]: See other join comments

File: openstack/cce/v3/addons/results.go [hunk 12]
```diff
@@ -29,7 +25,7 @@ type MetaData struct {
 	Annotations map[string]string `json:"annotaions"`
 }
 
-// Specifications to create an addon
+// Spec to create an addon
```
Comment (outcatcher) [inline openstack/cce/v3/addons/results.go:28]: To match both `start with name` and `natural reading` rules, you can put it in `Spec - specification to create an add-on`

File: openstack/cce/v3/addons/update.go [hunk 13]
```diff
@@ -0,0 +1,51 @@
...
+
+func Update(client *golangsdk.ServiceClient, id, clusterId string, opts UpdateOpts) (*Addon, error) {
+	b, err := golangsdk.BuildRequestBody(opts, "")
+	if err != nil {
+		return nil, err
+	}
+
+	raw, err := client.Put(fmt.Sprintf("https://%s.%s", clusterId, client.ResourceBaseURL()[8:])+
+		strings.Join([]string{"addons", id + "?cluster_id=" + clusterId}, "/"), b, nil, &golangsdk.RequestOpts{
+		OkCodes: []int{200},
```
Comment (outcatcher) [inline openstack/cce/v3/addons/update.go]: See other join comments

File: openstack/cce/v3/clusters/doc.go [hunk 14]
```diff
@@ -16,21 +16,21 @@ Example to List Clusters
...
+													       Version:"v1.7.3-r10",
+													       HostNetwork:clusters.HostNetworkSpec{VpcId:"3b9740a0-b44d-48f0-84ee-42eb166e54f7",
+																						SubnetId:"3e8e5957-649f-477b-9e5b-f1f75b21c045",},
+													       ContainerNetwork:clusters.ContainerNetworkSpec{Mode:"overlay_l2"},
+														},
+		         }
+	 	cluster,err := clusters.Create(client,createOpts).Extract()
+		if err != nil {
+			panic(err)
+		}
```
Comment (outcatcher) [inline openstack/cce/v3/clusters/doc.go:33]: Should this be really changed?

File: openstack/cce/v3/clusters/get.go [hunk 15]
```diff
@@ -0,0 +1,21 @@
...
+import (
+	"github.com/opentelekomcloud/gophertelekomcloud"
+	"github.com/opentelekomcloud/gophertelekomcloud/internal/extract"
+)
+
+// Get retrieves a particular cluster based on its unique ID.
+func Get(client *golangsdk.ServiceClient, id string) (*Clusters, error) {
+	raw, err := client.Get(client.ServiceURL("clusters", id), nil, &golangsdk.RequestOpts{
+		OkCodes:     []int{200},
+		MoreHeaders: RequestOpts, JSONBody: nil,
```
Comment (outcatcher) [inline openstack/cce/v3/clusters/get.go:12]: No need to fill empty values in structures

File: openstack/cce/v3/clusters/testing/requests_test.go [hunk 16]
```diff
@@ -5,24 +5,31 @@ import (
...
+	golangsdk "github.com/opentelekomcloud/gophertelekomcloud"
 	"github.com/opentelekomcloud/gophertelekomcloud/openstack/cce/v3/clusters"
-	fake "github.com/opentelekomcloud/gophertelekomcloud/openstack/cce/v3/common"
 	th "github.com/opentelekomcloud/gophertelekomcloud/testhelper"
+	"github.com/opentelekomcloud/gophertelekomcloud/testhelper/client"
 )
 
+func serviceClient() *golangsdk.ServiceClient {
+	sc := client.ServiceClient()
+	sc.ResourceBase = sc.Endpoint + "api/" + "v3/" + "projects/" + "c59fd21fd2a94963b822d8985b884673/"
```
Comment (outcatcher) [inline openstack/cce/v3/clusters/testing/requests_test.go:16]: Bit more readable to put it into `fmt.Sprintf`

File: openstack/cce/v3/nodepools/delete.go [hunk 17]
```diff
@@ -0,0 +1,12 @@
+package nodepools
+
+import "github.com/opentelekomcloud/gophertelekomcloud"
+
+// Delete will permanently delete a particular node pool based on its unique ID and cluster ID.
+func Delete(client *golangsdk.ServiceClient, clusterid, nodepoolid string) (err error) {
+	_, err = client.Delete(client.ServiceURL("clusters", clusterid, "nodepools", nodepoolid), &golangsdk.RequestOpts{
+		OkCodes:     []int{200},
+		MoreHeaders: RequestOpts, JSONBody: nil,
```
Comment (outcatcher) [inline openstack/cce/v3/nodepools/delete.go:9]: No need to set empty/default values in structures

File: openstack/cce/v3/nodepools/get.go [hunk 18]
```diff
@@ -0,0 +1,23 @@
...
+	"github.com/opentelekomcloud/gophertelekomcloud/internal/extract"
+)
+
+var RequestOpts = map[string]string{"Content-Type": "application/json"}
+
+// Get retrieves a particular node pool based on its unique ID and cluster ID.
+func Get(client *golangsdk.ServiceClient, clusterid, nodepoolid string) (*NodePool, error) {
+	raw, err := client.Get(client.ServiceURL("clusters", clusterid, "nodepools", nodepoolid), nil, &golangsdk.RequestOpts{
+		OkCodes:     []int{200},
+		MoreHeaders: RequestOpts, JSONBody: nil,
```
Comment (outcatcher) [inline openstack/cce/v3/nodepools/get.go:14]: No need to set empty/default values in structures

File: openstack/cce/v3/nodes/testing/requests_test.go [hunk 19]
```diff
@@ -5,18 +5,26 @@ import (
...
 	"github.com/opentelekomcloud/gophertelekomcloud/openstack/cce/v3/nodes"
 	th "github.com/opentelekomcloud/gophertelekomcloud/testhelper"
 )
 
+func serviceClient() *golangsdk.ServiceClient {
+	sc := client.ServiceClient()
+	sc.ResourceBase = sc.Endpoint + "api/" + "v3/" + "projects/" + "c59fd21fd2a94963b822d8985b884673/"
+	return sc
+}
+
```
Comment (outcatcher) [inline openstack/cce/v3/nodes/testing/requests_test.go:20]: Why have you copied `ServiceClient` to several packages?
They are literally the same function, as far as I can see.

**Review by outcatcher — CHANGES_REQUESTED** [review]

**Comment by Aloento** [comment]
Anthon-S: abot refactoring of cce. we should get rid v1 and use instead native k8s api

**Pull request closed without merging**